src/Controller/MakefileToJobTree.ts

```diff
@@ -0,0 +1,63 @@
+import { basename, dirname } from 'path'
+import { exec } from 'child_process'
+import { promisify } from 'util'
+import { Job } from '..'
+import { makefileTraceToJobTree } from './MakefileTraceToJobTree'
+import { MakeTracingError } from './MakeTracingError'
+
+/**
+ * Gets a trace from a makefile by running make with the --trace and --dry-run options.
+ *
+ * @param filePath - the path to the Makefile
+ * @returns the trace.
+ */
+async function getMakefileTrace(options: {
+	filePath?: string
+	targets?: string[]
+}): Promise<string> {
+	let command = 'make --trace --dry-run'
+
+	if (options.filePath) {
+		const fileName = basename(options.filePath)
+		const directoryName = dirname(options.filePath)
+
+		command += ` -f "${fileName}" --directory="${directoryName}"`
```

Comment:
And what if I want my fileName to be `"; sudo rm -rf /`?

Reply:
I switched from `exec` to `execFile()` which escapes its arguments.

Also, this is a relatively harmless security flaw since we're only running this on the local machine. If a user really wanted to input a malicious command, they could just do it themselves in terminal.